astropy/cosmology/io/tests/test_yaml.py

```diff
@@ -0,0 +1,155 @@
+    assert "node" in sig.parameters
+    assert sig.parameters["loader"].default is inspect._empty
+    assert sig.parameters["node"].default is inspect._empty
+
+    # it's too hard to manually construct a node, so we only test dump/load
+    # this is also a good round-trip test
+    yml = dump(Planck18)
+    cosmo = load(yml)
+    assert isinstance(cosmo, FlatLambdaCDM)
+    assert cosmo == Planck18
+    assert cosmo.meta == Planck18.meta
+
+
+##############################################################################
+# Test Unified I/O
+
+
+class ToFromYAMLTestMixin(IOTestMixinBase):
+    """
+    Tests for a Cosmology[To/From]Format with ``format="yaml"``.
+    This class will not be directly called by :mod:`pytest` since its name does
+    not begin with ``Test``. To activate the contained tests this class must
+    be inherited in a subclass. Subclasses must define a :func:`pytest.fixture`
+    ``cosmo`` that returns/yields an instance of a |Cosmology|.
+    See ``TestCosmologyToFromFormat`` or ``TestCosmology`` for examples.
+    """
+
+    def test_to_from_yaml_instance(self, cosmo, to_format, from_format):
+        """Test cosmology -> YAML -> cosmology."""
+        # ------------
+        # To YAML
+
+        yml = to_format('yaml')
+        assert isinstance(yml, str)  # test type
+        assert yml.startswith("!astropy.cosmology.")
+
+        # ------------
+        # From YAML
+
+        # tests are different if the last argument is a **kwarg
+        if tuple(cosmo._init_signature.parameters.values())[-1].kind == 4:
+            got = from_format(yml, format="yaml")
+
+            assert got.name == cosmo.name
+            assert got.meta == cosmo.meta
+
+            return  # don't continue testing
+
+        # it won't error if everything matches up
+        got = from_format(yml, format="yaml")
+        assert got == cosmo
+        assert got.meta == cosmo.meta
+
+        # it auto-identifies 'format'
+        got = from_format(yml)
+        assert got == cosmo
+        assert got.meta == cosmo.meta
+
+    # # TODO! this is a challenging test to write. It's also unlikely to happen.
+    # def test_fromformat_subclass_partial_info_yaml(self, cosmo):
+    #     """
+    #     Test writing from an instance and reading from that class.
+    #     This works with missing information.
+    #     """
+
+
+class TestToFromYAML(IOFormatTestBase, ToFromYAMLTestMixin):
+    """
+    Directly test ``to/from_yaml``.
+    These are not public API and are discouraged from use, in favor of
+    ``Cosmology.to/from_format(..., format="yaml")``, but should be tested
+    regardless b/c 3rd party packages might use these in their Cosmology I/O.
+    Also, it's cheap to test.
+    """
+
+    def setup_class(self):
+        self.functions = {"to": to_yaml, "from": from_yaml}
+
+    @pytest.fixture(scope="class", autouse=True)
+    def setup(self):
+        """Setup and teardown for tests."""
+        yield  # run tests
```

Comment:
@nstarman - I see that you and @mhvk are big fans of layers of abstraction and layers of fixtures in testing. I will just say that it makes it quite a lot harder for independent reviewers to understand what the tests are actually doing, and so I'm not that enthusiastic about the approach. I've had a hard time understanding the tests from @mhvk and same applies here.

Simple is better than complex. Having the final "test" of a feature boil down to one or two lines may seem  elegant, but I cannot understand the test code directly without reading and understanding the upstream base classes and fixtures.

What is different in these tests from simply looping over each of the built-in cosmology classes and doing `dump` `load` `to_format` and `from_format` tests? This might avoid or simplify the special-case handling of `if tuple(cosmo._init_signature.parameters.values())[-1].kind == 4:`.

Reply:
Haha. Guilty as charged. Apologies for that.
This test suite is structured so that 
1. Descriptors on Cosmology, like ``meta/read/write/from_format/to_format`` are parents to ``TestCosmology``so that all tests for the descriptor are run on every Cosmology subclass, like ``TestFLRW`` and ``TestFlatLambdaCDM``
2. For the I/O descriptors the tests for each format, like "yaml", are parents to the descriptor test so that the format tests are run for each Cosmology subclass.

So the inheritance looks like is ``TestFLRW(TestCosmology(ToFromFormatTestMixin(ToFromMappingTestMixin , ToFromYAMLTestMixin)))``
Where ``ToFromFormatTestMixin`` tests the descriptors ``to/from_format`` and is fed the tests from each registered format.

> What is different in these tests from simply looping over each of the built-in cosmology classes and doing dump load to_format and from_format tests? This might avoid or simplify the special-case handling of if tuple(cosmo._init_signature.parameters.values())[-1].kind == 4:.

Most of these tests are just looping over each of the built-in cosmology classes. But no built-in cosmology has a kwargs in ``__init__`` (that's the ``tuple(cosmo._init_signature.parameters.values())[-1].kind == 4``).
I test this case because a fair number of things were broken in non-intuitive ways, like ``clone`` and I/O with metadata, especially when there's partial information allowed like a cosmology <-> dict.  I added the ``kwargs`` tests a number of PRs ago and fixed things to make it work. Having kwargs necessitates a slightly different treatment of metadata, so it needs to be special cased.
Except I don't think this is the case for ``yaml`` unless someone is editing the YAML strings and adding extra args/kwargs! I have a TODO! on the partial information test because I think this quite unlikely. If someone ever hand-assembles a valid Cosmology yaml string, then I'll write the test!
To end a long explanation, ``tuple(cosmo._init_signature.parameters.values())[-1].kind == 4`` is not necessary for this test. I'm removing the check.